libs/src/services/discoveryProvider/DiscoveryProviderSelection.js

```diff
@@ -0,0 +1,213 @@
+const ServiceSelection = require('../../service-selection/ServiceSelection')
+const {
+  DISCOVERY_PROVIDER_TIMESTAMP,
+  DISCOVERY_SERVICE_NAME,
+  UNHEALTHY_BLOCK_DIFF,
+  DISCOVERY_PROVIDER_RESELECT_TIMEOUT,
+  REGRESSED_MODE_TIMEOUT
+} = require('./constants')
+const semver = require('semver')
+
+const PREVIOUS_VERSIONS_TO_CHECK = 5
+
+let localStorage
+if (typeof window === 'undefined' || window === null) {
+  const LocalStorage = require('node-localstorage').LocalStorage
+  localStorage = new LocalStorage('./local-storage')
+} else {
+  localStorage = window.localStorage
+}
+
+class DiscoveryProviderSelection extends ServiceSelection {
+  constructor (config, ethContracts) {
+    super({
+      /**
+       * Gets the "current" expected service version as well as
+       * the list of registered providers from chain
+       */
+      getServices: async () => {
+        this.currentVersion = await ethContracts.getCurrentVersion(DISCOVERY_SERVICE_NAME)
+        return this.ethContracts.getServiceProviderList(DISCOVERY_SERVICE_NAME)
+      },
+      ...config
+    })
+    this.ethContracts = ethContracts
+    this.currentVersion = null
+
+    // Whether or not we are running in `regressed` mode, meaning we were
+    // unable to select a discovery provider that was up-to-date. Clients may
+    // want to consider blocking writes.
+    this._regressedMode = false
+  
+    // Set of valid past discovery provider versions registered on chain
+    this.validVersions = null
+  }
+
+  /** Retrieves a cached discovery provider from localstorage */
+  getCached () {
+    if (localStorage) {
+      const discProvTimestamp = localStorage.getItem(DISCOVERY_PROVIDER_TIMESTAMP)
+      if (discProvTimestamp) {
+        const { endpoint: latestEndpoint, timestamp } = JSON.parse(discProvTimestamp)
+        const inWhitelist = !this.whitelist || this.whitelist.has(latestEndpoint)
```

Comment:
so if `this.whitelist` is null, the cached dp is going to be considered in the whitelist?

Reply:
if whitelist is null, the cached dp is valid.
it's not in the whitelist because there is no whitelist ;)